### DISCUSSION TITLE
Cannot upload large files via AWS S3 SDK

### DISCUSSION THREAD

**Discussion by bryanyip94** [minio/minio#13332] [Q&A] [answered]
I am trying to upload a file to my Minio bucket via the python AWS S3 SDK functionality. For small files this works fine. However for larger files (~>16mb, where multipart upload is needed), this does not work, or extremely slow.

## Expected Behavior
For files larger than 16mb, the upload functionality from python AWS S3 SDK should work. Via the Minio SDK it works fine, but in our architecture, using the AWS S3 SDK is preferred. 

## Current Behavior
The upload either gets stuck, nothing happens for a long while (several hours for 16mb...). Or it gets uploaded, but with an upload speed of 16mn per 30min. 

## Steps to Reproduce (for bugs)
`import boto3

clientlocal = boto3.client('s3',
                  endpoint_url="http://localhost:9000",
                  aws_access_key_id='minioadmin',
                  aws_secret_access_key='minioadmin')
clientlocal.upload_file('file.zip', Bucket=bucketName, Key='file.zip')
`

In this example file can be zip or any large file. The file I used was 70mb.

## Your Environment
I used the latest Minio version (RELEASE.2021-09-18T18-09-59Z), but have also tried several other version. 

I both tried on an environment created with yaml files and 1 by running the following command:
docker run \
  -p 9000:9000 \
  -p 9001:9001 \
  minio/minio server /data --console-address ":9001"

**Comment by harshavardhana** [comment]
@bryanyip94 can you share `mc admin trace -v` output when you are uploading > 16MiB files?

**Comment by bryanyip94** [comment]
> @bryanyip94 can you share `mc admin trace -v` output when you are uploading > 16MiB files?

Hereby a screenshot, and also here it gets stuck.

![image](https://user-images.githubusercontent.com/82947941/134411407-585f00de-cf70-4a39-9205-68724ebe2203.png)

**Comment by harshavardhana** [comment]
> > @bryanyip94 can you share `mc admin trace -v` output when you are uploading > 16MiB files?
> 
> Hereby a screenshot, and also here it gets stuck.
> 
> ![image](https://user-images.githubusercontent.com/82947941/134411407-585f00de-cf70-4a39-9205-68724ebe2203.png)

that doesn't make sense @bryanyip94 it looks like `boto3` is not uploading anything - can you check `curl https://minio:port/minio/v2/metrics/cluster | grep inflight`

**Comment by bryanyip94** [comment]
> curl https://minio:port/minio/v2/metrics/cluster | grep inflight

![image](https://user-images.githubusercontent.com/82947941/134414696-1bc5f6ba-6aed-4a2e-9d1c-0a2a3ff817ca.png)

And if cancel my upload request, so stop the python script, I get the following logs, which means the script was actually still running:

![image](https://user-images.githubusercontent.com/82947941/134414800-9aface66-61bd-441d-8f6c-7b7902651a21.png)

And using exactly the same script, but uploading a small jpg file, I get the following logs:

![image](https://user-images.githubusercontent.com/82947941/134415026-29ed239d-a70a-402f-b504-e0f76fd136bf.png)

**Comment by harshavardhana** [comment]
> > curl https://minio:port/minio/v2/metrics/cluster | grep inflight
> 
> ![image](https://user-images.githubusercontent.com/82947941/134414696-1bc5f6ba-6aed-4a2e-9d1c-0a2a3ff817ca.png)
> 
> And if cancel my upload request, so stop the python script, I get the following logs, which means the script was actually still running:
> 
> ![image](https://user-images.githubusercontent.com/82947941/134414800-9aface66-61bd-441d-8f6c-7b7902651a21.png)
> 
> And using exactly the same script, but uploading a small jpg file, I get the following logs:
> 
> ![image](https://user-images.githubusercontent.com/82947941/134415026-29ed239d-a70a-402f-b504-e0f76fd136bf.png)

Prometheus endpoint is not public by default, I thought you know this. Please make sure to enable Prometheus endpoint as public to run the command that I requested.

**Comment by harshavardhana** [comment]
Also can you use boto3 on Linux, is your MinIO also running on windows?

**Comment by bryanyip94** [comment]
> curl https://minio:port/minio/v2/metrics/cluster | grep inflight

We have now enabled Prometheus on our MinIO, the one running on our Linux Kubernetes cluster.

The last line of the screenshot below is added when I try to upload a file (putobjectpart).

![image](https://user-images.githubusercontent.com/82947941/134516138-a79d4395-5942-4463-9337-4ce606ed2b1c.png)


Also after waiting for a while, I am getting the following error:

ReadTimeoutError: Read timeout on endpoint URL: "http://minio-aws-dev.vpd-acn.com:9000/vpd-qa-visualprocessdiscovery-eu-central-1/Preprocess.zip?uploadId=1875104c-5972-4b62-aa6f-c9e9454f61cc&partNumber=8"

**Comment by kannappanr** [comment]
I tried the given Python file against minio created using the given docker command and minio created with stand alone FS mode and locally erasure coded setup. I could not reproduce the issue.

**Comment by harshavardhana** [comment]
> I tried the given Python file against minio created using the given docker command and minio created with stand alone FS mode and locally erasure coded setup. I could not reproduce the issue.

I suspect there is a disk issue here @kannappanr

**Comment by klauspost** [comment]
`ReadTimeoutError` also indicates a client side timeout, it is not something that is caused by the server.

**Comment by bryanyip94** [comment]
So on my local pc, it works now. Instead of running 

docker run
-p 9000:9000
-p 9001:9001
minio/minio server /data --console-address ":9001"

I know deployed minio in my minikube with pvc volume. There it uploaded a file of 140mb in 4sec.

However, afterwards I used the same yaml files to deploy to my kubernetes cluster in AWS. If I do port-forward here to my minio pod, it still uploads my file, but the same file takes 50sec. But the real problem comes when I attach our loadbalancer to the deployment. Then the upload just does not work for large files. This you can see in the screenshot (of curl https://minio:port/minio/v2/metrics/cluster | grep inflight) in the previous comment.

**Comment by bryanyip94** [comment]
> `ReadTimeoutError` also indicates a client side timeout, it is not something that is caused by the server.

Found this out, there is a timeout set on our load balancer. Uploading 27mb takes 48minutes if we don't have a limit... which for me sounds too long.

Just wondering is MinIO actually fully compatible with load balancers like AWS elb?

**Comment by harshavardhana** [comment]
> > `ReadTimeoutError` also indicates a client side timeout, it is not something that is caused by the server.
> 
> Found this out, there is a timeout set on our load balancer. Uploading 27mb takes 48minutes if we don't have a limit... which for me sounds too long.
> 
> Just wondering is MinIO actually fully compatible with load balancers like AWS elb?

We don't do anything special, you have to make sure to upload from a faster network here.

**Comment by bryanyip94** [comment]
> > > `ReadTimeoutError` also indicates a client side timeout, it is not something that is caused by the server.
> > 
> > 
> > Found this out, there is a timeout set on our load balancer. Uploading 27mb takes 48minutes if we don't have a limit... which for me sounds too long.
> > Just wondering is MinIO actually fully compatible with load balancers like AWS elb?
> 
> We don't do anything special, you have to make sure to upload from a faster network here.

But how come port-forward from pod works and uploads with an okay speed, but the aws elb load balancer attached to the pod does not? Neither does the load balancer from gcp work.

**Comment by harshavardhana** [comment]
> But how come port-forward from pod works and uploads with an okay speed, but the aws elb load balancer attached to the pod does not? Neither does the load balancer from gcp work.

Do not know really @bryanyip94 as long as MinIO is taking uploads this is not a problem on our end.

**Comment by bryanyip94** [comment]
> > But how come port-forward from pod works and uploads with an okay speed, but the aws elb load balancer attached to the pod does not? Neither does the load balancer from gcp work.
> 
> Do not know really @bryanyip94 as long as MinIO is taking uploads this is not a problem on our end.

Okay but somehow MinIO is not compatible with AWS ELB when wanting to upload files it seems. 

How do you usually upload large files when MinIO is deployed in the cloud? Is there another way than using AWS ELB?

**Comment by harshavardhana** [comment]
> How do you usually upload large files when MinIO is deployed in the cloud? Is there another way than using AWS ELB?

We use AWS ELB and haven't seen such issues.

**Comment by bryanyip94** [comment, marked as answer]
I finally found a solution for the issue. The file upload is working extremely slow on http, but fast on https. Not sure why, but switching to https helps.
